### PR TITLE
[InstCombine] Use KnownBits predicate helpers

### PR DESCRIPTION
**Pull request by nikic**
Inside foldICmpUsingKnownBits(), instead of rolling our own logic based on min/max values, make use of ICmpInst::compare() working on KnownBits. This gives better results for the equality predicates. In practice, the improvement is only for pointers, because isKnownNonEqual() handles the non-pointer case.

I've adjusted some tests to prevent the new fold from triggering, to retain their original intent of testing constant expressions.

### CODE REVIEW

**Comment by nikic** [comment]
This may have a bit of compile-time impact (it's close to the noise threshold): https://llvm-compile-time-tracker.com/compare.php?from=36f21eedcfd06ff97ead9625adbf6d8153edd233&to=a371228364bfdec8864e4a3af655d45c8700a87f&stat=instructions:u

If preferred I could only use `KnownBits::eq` and leave the rest of the code alone. I figured it would be nice to consolidate the handling though.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -6544,6 +6544,35 @@ bool InstCombinerImpl::replacedSelectWithOperand(SelectInst *SI,
   return false;
 }
 
+static std::optional<bool> compareKnownBits(ICmpInst::Predicate Pred,
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6547]: We already implemented similar logic in `CombinerHelper::matchICmpToTrueFalseKnownBits`. Do we need a common helper?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:6547]: Opened https://github.com/llvm/llvm-project/pull/115878 to add the common helper.

**Comment by nikic** [comment]
The reason this only makes a different for pointers is that simplifyICmpInst calls isKnownNonEqual, which does a KnownBits based check for integers, but not pointers.

So an alternative fix would be to adjust isKnownNonEqual() to handle this for pointers as well.

**Comment by goldsteinn** [comment]
> The reason this only makes a different for pointers is that simplifyICmpInst calls isKnownNonEqual, which does a KnownBits based check for integers, but not pointers.
> 
> So an alternative fix would be to adjust isKnownNonEqual() to handle this for pointers as well.

The two don't seem mutually exclusive. This patch has value IMO even if it's just dropping some bespoke logic for better tested/implemented apis

**Comment by nikic** [comment]
> > The reason this only makes a different for pointers is that simplifyICmpInst calls isKnownNonEqual, which does a KnownBits based check for integers, but not pointers.
> > So an alternative fix would be to adjust isKnownNonEqual() to handle this for pointers as well.
> 
> The two don't seem mutually exclusive. This patch has value IMO even if it's just dropping some bespoke logic for better tested/implemented apis

Generally agree. However, I just tested the isKnownNonEqual change, and that one has more significant compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=6d8d9fc8d279623cca94b2b875a92517ed308f18&to=49f63dc5015abe32735e99c39cf86bab32a7a19a&stat=instructions:u It's adding 0.2% to clang thin link.

So I'm inclined to go with this patch only.

**Comment by goldsteinn** [comment]
> > > The reason this only makes a different for pointers is that simplifyICmpInst calls isKnownNonEqual, which does a KnownBits based check for integers, but not pointers.
> > > So an alternative fix would be to adjust isKnownNonEqual() to handle this for pointers as well.
> > 
> > 
> > The two don't seem mutually exclusive. This patch has value IMO even if it's just dropping some bespoke logic for better tested/implemented apis
> 
> Generally agree. However, I just tested the isKnownNonEqual change, and that one has more significant compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=6d8d9fc8d279623cca94b2b875a92517ed308f18&to=49f63dc5015abe32735e99c39cf86bab32a7a19a&stat=instructions:u It's adding 0.2% to clang thin link.
> 
> So I'm inclined to go with this patch only.

Fair enough, is there no other independent value to IsKnownNonEqual for ptrs?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Now using the common helper.

**Review by dtcxzyw — APPROVED** [review]
LG

**Pull request merged**